// natis/dist/helper.js
require('mstr');
var dateutil = require('mdate');
var strs = "abcdefghijklmnopqrstuvwxyz0123456789abc";

module.exports = function(template){

    //['字段值']  {{ val | q}}
	template.helper('q',function(val){
        if(val === 0 || val ==='0'){
            return "'0'";
        }
		return "'" + escape(val || "") + "'";
	});

    //[字段名 = '字段值'] {{ val | eq:'col'}}
    template.helper('eq',function(val,col,blank){
        var rst = val;
        if(rst == '' && !blank){
            return '';
        }
        if(val === 0){
            rst = "0";
        }
       return " " + col + " = '" + escape(rst || "") + "' ";
    });

    //['%字段值%'] {{ val | p}}
	template.helper('p',function(val){
		return "'%" + escape(val || "") + "%'";
	});

    //[,] {{data | m:index}}
    template.helper('m',function(data,index){
        if(data.length > (index+1)){
            return ',';
        }
        return '';
    });

    //['当前时间'] {{now | d}}
    template.helper('d',function(val,format){
        var result = dateutil.format(val,format);
        return "'" + (result || "") + "'";
    });

    //[字段名 = '当前日期'] {{now | ed:'col'}}
    template.helper('ed',function(val,col){
        return " " +col + " = '" + dateutil.format() + "' ";
    });


    //[,字段名 = '字段值'] {{val | set:'col'}}
    template.helper('set',function(val,col){
        if(val == null){
            return '';
        }
        return "," + col + " = '" + escape(val) + "' ";
    });

    //[ and 字段名 in (字段值)] {{val | in:'col'}}
    template.helper('in',function(val,col){
        if(val == null || val === '' || val === '-'){
            return '';
        }
        return ' and ' + col + ' in (' + escape(val) + ') '
    });

    //[ and 字段名 = '字段值']    {{val | and:col,'eq'}}
    //[ and 字段名 > '字段值'] {{val | and:col,'gt'}}
    //[ and 字段名 >= '字段值'] {{val | and:col,'gte'}}
    //[ and 字段名 < '字段值'] {{val | and:col,'lt'}}
    //[ and 字段名 <= '字段值'] {{val | and:col,'lte'}}
    //[ and 字段名 != '字段值'] {{val | and:col,'neq'}}
    //[ and 字段名 like '%字段值%'] {{val | and:col,'like'}}
    //[ and 字段名 in (1,2)] {{val | and:col,'in'}}
    //[ and (字段1 like '%值%' or 字段2 like '%值%' ...)] {{val | and:'col1,col2...','like'}}

    template.helper('and',function(val,col,tag,pattern) {
        pattern = pattern || 'digit';
        if(val == null) {
             return '';
        };
        if(typeof val == "string"){
            val = val.trim();
        }
        if(val === '' || val == '-'){
            return ''
        }
        switch (pattern){
            case 'digit' :
                if(val === 0 || val === -1 || val === "0" || val === "-1"){
                    return '';
                }
                break;
        }
        tag = tag || 'eq';

        var tags = {
            eq : '=',
            neq : '!=',
            lt : '<',
            lte : '<=',
            gt : '>',
            gte : '>=',
            like : 'like',
            in : 'in'
        }
        if(tag == 'like'){
            val = "%" + escape(val) + "%";
        }
        else if(tag == 'in'){
            val = " (" + val + ") ";
        }
        var realTag = tags[tag];
        if(realTag){
            //同一个值被多个字段引用
            var cols = col.split(",");
            if(cols.length == 1){
                return " and " + getSql(col,realTag,val);
            }
            else{
                var sqls = [];
                for(var i =0;i<cols.length;i++){
                    sqls.push(getSql(cols[i],realTag,val));
                }
                return " and (" + sqls.join(" or ") + " )";
            }
        }
        else{
            return '';
        }
    });

    // [ order by id desc ] {{ val | orderby}}
    template.helper('orderBy',function(val,defaultVal){
        if(val == null || val == ''){
            return ' order by ' + (defaultVal || 'id asc') + ' ';
        }
        return ' order by ' + val + ' ';
    });
    // [limit offset,pageSize] {{offset | limit:pageSize}}
    template.helper('limit', function(offset,pageSize){
        if(pageSize > 0){
            return ' limit ' + offset + ',' + pageSize;
        }
        return '';
    });

    function getSql(col,tag,val){
        return " " + col + " " + tag  + " '" + escape(val) + "' ";
    }

    function escape(val){
        val = val + '';
        return val.replace(/'/g,"\\'");
    }
}
